File: 프로그래머스/스택큐/올바른괄호.py
#https://school.programmers.co.kr/learn/courses/30/lessons/12909

def solution(s):
    stack = []
    
    for ch in s:
        if ch == '(': #여는 괄호라면 스택에 넣음
            stack.append(ch)
        elif ch == ')': #닫는 괄호라면 스택의 마지막 괄호와 짝이 맞는지 비교
            if not stack: #스택이 비어있는 경우
                return False
            
            top = stack.pop() #스택에서 꺼냄
            #꺼낸게 이게 아니라면 짝 안맞는 거
            if top != '(' :
                return False

    #모든 순회가 끝나면 스택 비어있어야함
    return len(stack) == 0

File: 프로그래머스/KB/올바른괄호.java
// https://school.programmers.co.kr/learn/courses/30/lessons/12909

class Solution {
    boolean solution(String s) {
        int open = 0;
        int len = s.length();

        //괄호는 짝수여야함 (짝이 맞아야하기 때문)
        if (len % 2 != 0) return false;

        //열린괄호면 ++, 닫힌괄호 만나면 -- > 최종적으로 0이 되야함.
        for (int i = 0; i < len; i++) {
            if (s.charAt(i) == '(') {
                open++;
            } else {
                open--;
            }
            //닫힌괄호 먼저 나오면 음수가 되니까 false
            if (open < 0) return false;
        }

        return open == 0;
    }
}
